File: src/main/java/cpsc224/pools/Pool.java
package cpsc224.pools;

import java.util.*;

/**
 * A class to represent a pool of possible objects that can be created.
 * Each object is assigned a weight that determines its chance of being selected.
 * @param <T> the type of objects in the pool
 */
public abstract class Pool<T> {

    Random rand = new Random();
    protected HashMap<ObjectCreator<T>, Double> objectCreators = new HashMap<>();
    protected double totalWeight;

    /**
     * Creates a new pool.
     */
    public Pool() {
        totalWeight = 0.0;
    }

    /**
     * Adds an object creator to the pool.
     * @param objectCreator the object creator
     * @param weight determines the chance of being chosen
     */
    public void addObjectCreator(ObjectCreator<T> objectCreator, double weight) {
        Double prevWeight = objectCreators.put(objectCreator, weight);

        // subtract the prev weight if this object creator already mapped to a value
        if (prevWeight != null)
            totalWeight -= prevWeight;

        // increase total weight
        totalWeight += weight;
    }

    /**
     * Creates a new object from the pool.
     * @return the new object
     */
    public T createNew() {
        double randValue = rand.nextDouble() * totalWeight;
        double cumulativeWeight = 0.0;

        for (ObjectCreator<T> itemCreator : objectCreators.keySet()) {
            cumulativeWeight += objectCreators.get(itemCreator);
            if (randValue <= cumulativeWeight) {
                return itemCreator.createItem();
            }
        }
        return null;
    }
}

File: web/src/game/pools/Pool.ts
export type ObjectCreator<T> = () => T;

export class Pool<T> {
  private creators = new Map<ObjectCreator<T>, number>();
  private totalWeight = 0;

  addObjectCreator(creator: ObjectCreator<T>, weight: number): void {
    const prev = this.creators.get(creator);
    if (prev !== undefined) this.totalWeight -= prev;
    this.creators.set(creator, weight);
    this.totalWeight += weight;
  }

  getCreator(): ObjectCreator<T> {
    let roll = Math.random() * this.totalWeight;
    for (const [creator, weight] of this.creators) {
      roll -= weight;
      if (roll <= 0) return creator;
    }
    return [...this.creators.keys()][0];
  }

  createNew(): T {
    return this.getCreator()();
  }
}
